File: Proyecto Mis Finanzas Python/src/gestorAplicación/interno/cuenta.py
from abc import ABC, abstractmethod
from datetime import date, datetime
from gestorAplicación.externo.banco import Banco
import math


class Cuenta(ABC):
    #Atributos de clase
    _cuentasTotales = []

    #Constructor
    def __init__(self, _banco = Banco(), **kwargs):
        self._clave = None
        self._nombre = ""
        self._divisa = None
        self._titular = None
        self._banco = None
        #Atributos de instancia
        Cuenta._cuentasTotales.append(self)
        self._id = len(Cuenta._cuentasTotales)
        for key in kwargs:
            if key == "banco":
                self._banco = kwargs[key]
            if key == "clave":
                self._clave = kwargs[key]
            if key == "nombre":
                self._nombre = kwargs[key]
            if key == "divisa":
                self._divisa = kwargs[key]

    @abstractmethod
    def crearCuenta(self, banco, clave, nombre, **kwargs):
        pass
    #El kwargs hace referencia a la sobrecarga de la divisa

    def __str__(self):
        return "Cuenta: " + self._nombre + "\n# " + self._id + "\nTitular: " + self.getTitular().getNombre() + "\nBanco: " + self._banco.getNombre() + "\nDivisa: " + self._divisa

    # Métodos para la funcionalidad asesoramiento de inversiones
    def gota_gota(cantidadPrestamo, user, gota):
        from .movimientos import Movimientos
        from .categoria import Categoria
        from .ahorros import Ahorros
        mayor = 0
        contador = 0
        for i in range(len(user.getCuentasAsociadas())):
            if isinstance(user.getCuentasAsociadas()[i], Ahorros):
                if user.getCuentasAsociadas()[i].getSaldo() > mayor:
                    mayor = user.getCuentasAsociadas()[i].getSaldo()
                    contador = i
                else:
                    continue
            else:
                if user.getCuentasAsociadas()[i].getDisponible() > mayor:
                    mayor = user.getCuentasAsociadas()[i].getDisponible()
                    contador = i
            
            movimiento = Movimientos(destino = user.getCuentasAsociadas()[contador], cantidad = cantidadPrestamo, categoria = Categoria.OTROS, fecha = datetime.now())
            Movimientos.getMovimientosTotales().remove(movimiento)
            return user.getCuentasAsociadas()[contador]
        
    def vaciar_cuenta(gota):
        pass

    #Métodos para funcionalidad de cambio de divisa
    @staticmethod
    def hacer_cambio(escogencia, monto, destino, user, exacto=False):
        from .movimientos import Movimientos
        origen = escogencia.get_origen()
        if exacto:
            pagar = monto / (1 - escogencia.get_banco().get_estado_asociado().get_tasa_impuestos())
            pagar /= (1 - escogencia.get_couta_manejo())
            pagar /= escogencia.get_cantidad()
            m = Movimientos(escogencia.get_banco(), origen, destino, escogencia.get_divisa(), escogencia.get_divisa_aux(), escogencia.get_couta_manejo(), pagar, datetime.datetime.now())
            origen.set_saldo(origen.get_saldo() - pagar)
            destino.set_saldo(destino.get_saldo() + monto)
        else:
            cambiado = monto * (1 - escogencia.get_banco().get_estado_asociado().get_tasa_impuestos())
            cambiado *= (1 - escogencia.get_couta_manejo())
            cambiado *= escogencia.get_cantidad()
            m = Movimientos(escogencia.get_banco(), origen, destino, escogencia.get_divisa(), escogencia.get_divisa_aux(), escogencia.get_couta_manejo(), monto, datetime.datetime.now())
            origen.set_saldo(origen.get_saldo() - monto)
            destino.set_saldo(destino.get_saldo() + cambiado)

        user.asociar_movimiento(m)
        for i in range(len(user.get_bancos_asociados())):
            user.get_bancos_asociados()[i].set_asociado(False)

    @staticmethod
    def comprobar_saldo(cuenta, monto=None):
        ahorro = cuenta
        if monto is None:
            return ahorro.get_saldo() >= monto
        else:
            pagar = monto / (1 - cuenta.get_banco().get_estado_asociado().get_tasa_impuestos())
            pagar /= (1 - cuenta.get_couta_manejo())
            pagar /= cuenta.get_cantidad()
            return ahorro.get_saldo() >= pagar

    @staticmethod
    def obtener_cuentas_divisa(usuario, divisa):
        cuentas_b = []
        for ahorro in usuario.get_cuentas_ahorros_asociadas():
            if ahorro.get_divisa() == divisa:
                cuentas_b.append(ahorro)
        return cuentas_b


    #Realizar el método del compare to
    def compareTo(self, cuenta):
        if self.getId() > cuenta.getId():
            return 1
        elif self.getId() < cuenta.getId():
            return -1
        else:
            return 0

    #¿Se maneja ligadura dinámica?
    def invertirSaldo(self):
        pass

    def retornoCuotaMensual(self, deudaActual):
        banco = self.getBanco()
        cuotaMensual = []
        if (banco.getComision() + banco.getEstadoAsociado().getInteres_bancario_corriente()) < 1:
            #Cuota del estado y del banco
            cuota1 = deudaActual*banco.getComision() + banco.getEstadoAsociado().getIntereses_bancario_corriente()*deudaActual
            cuota2 = (deudaActual - cuota1) / 2
            cuotaMensual[0] = deudaActual*banco.getComision() + banco.getEstadoAsociado().getInteres_bancario_corriente()*deudaActual
            cuotaMensual[1] = cuota2
            cuotaMensual[2] = cuota2
        else:
            cuotaMensual[0] = deudaActual / 3
            cuotaMensual[1] = deudaActual / 3
            cuotaMensual[2] = deudaActual / 3
        return cuotaMensual

    #Realizar el método equals
    def equals(self, cuenta):
        if self.getId() == cuenta.getId():
            return True
        else:
            return False

    @staticmethod
    def limpiarPropiedades(arreglo):
        arreglo.remove("cuentasTotales")
        #Verificar que otras variables se crean
    #Verificar su uso

    @staticmethod
    def dineroATenerDisponible(cuenta, divisaB):
        from .movimientos import Movimientos
        deuda = cuenta.getCupo() - cuenta.getDisponible()
        if cuenta.getDivisa() == divisaB:
            print(1)
            return cuenta.getCupo() - cuenta.getDisponible()
        cambio_div = Movimientos(divisa=cuenta.getDivisa(), divisaAux=divisaB, owner = cuenta.getTitular())
        existe_cambio = Movimientos.facilitar_informacion(cambio_div)
        if len(existe_cambio) == 0:
            return 0
        
        cuentas_posibles = []
        for conta in cuenta.getTitular().getCuentasCorrienteAsociadas():
            if conta.getDivisa() == cuenta.getDivisa():
                cuentas_posibles.append(conta)
        
        cadena = cuenta.getDivisa().name + divisaB.name
        imprimir = Banco.cotizar_taza_aux(cuenta.getTitular(), existe_cambio, cadena, cuentas_posibles)
        cambio_max = 0
        valor = 999999999
        for cotizacion in imprimir:
            valor = cotizacion.getCantidad() / ((1-cotizacion.getCuotaManejo()) * (1 - cotizacion.getBanco().getEstadoAsociado().getTasa_impuestos()))
            if valor > cambio_max:
                cambio_max = valor

        return cambio_max * deuda

    def __str__(self):
        return "Cuenta: " + self._nombre + "\n# " + self._id + "\nTitular: " + self.getTitular.getNombre() + "\nBanco: " + self._banco.getNombre() + "\nDivisa: " + self._divisa

    #Métodos Get & Set
    @classmethod
    def getCuentasTotales(cls):
        return cls._cuentasTotales
    @classmethod
    def setCuentasTotales(cls, cuentasTotales):
        cls._cuentasTotales = cuentasTotales

    def getTitular(self):
        return self._titular
    def setTitular(self, titular):
        self._titular = titular

    def getDivisa(self):
        return self._divisa
    def setDivisa(self, divisa):
        self._divisa = divisa

    def getNombre(self):
        return self._nombre
    def setNombre(self, nombre):
        self._nombre = nombre

    def getId(self):
        return self._id
    def setId(self, id):
        self._id = id

    def getBanco(self):
        return self._banco
    def setBanco(self, banco):
        self._banco = banco

    def getClave(self):
        return self._clave
    def setClave(self, clave):
        self._clave = clave

File: Proyecto Mis Finanzas Java/src/gestorAplicación/interno/Cuenta.java
package gestorAplicación.interno;

import java.util.ArrayList;
import java.util.Date;
import gestorAplicación.externo.Banco;
import gestorAplicación.externo.Divisas;
import gestorAplicación.externo.Estado;
import gestorAplicación.externo.Tablas;

import java.io.Serializable;
import java.lang.reflect.Field;
import java.time.Instant;

public abstract class Cuenta implements Serializable, Comparable<Cuenta>{
	//Atributos
	private static final long serialVersionUID = 4L;
	public static final String nombreD = "Cuentas";
	private Usuario titular;
	private int clave;
	protected Divisas divisa;
	protected String nombre;
	protected int id;
	protected Banco banco;
	private static transient ArrayList<Cuenta> cuentasTotales = new ArrayList<Cuenta>();

	
	//Constructores
	protected Cuenta(Banco banco, int clave, Divisas divisa, String nombre) {
		this.clave = clave;
		this.divisa = divisa;
		this.nombre = nombre;
		this.banco = banco;	
		cuentasTotales.add(this);
		this.setId(cuentasTotales.size());
	}	
	
	protected Cuenta(Banco banco, int clave, String nombre) {;
		this.clave = clave;
		//Acceder a la divisa definida como predeterminada por el banco
		this.divisa = banco.getEstadoAsociado().getDivisa();
		this.nombre = nombre;
		this.banco = banco;
		cuentasTotales.add(this);
		this.setId(cuentasTotales.size());
	}
	
	protected Cuenta() {
		cuentasTotales.add(this);
		this.setId(cuentasTotales.size());
	}
	
	//Métodos
	public abstract Cuenta crearCuenta(Banco banco, int clave, Divisas divisa, String nombre);
	
	public abstract Cuenta crearCuenta(Banco banco, int clave, String nombre);
	
	// Funcionalidad Asesor de Inversiones
	public static Cuenta gotaGota(double cantidadPrestamo, Usuario user, Ahorros gota) {

		double mayor = 0;
		int contador = 0;
		if (user.getCuentasAhorrosAsociadas().size() != 0) {
			for (int i = 0; i < user.getCuentasAhorrosAsociadas().size(); i++) {
				if (user.getCuentasAhorrosAsociadas().get(i).getSaldo() > mayor) {
					mayor = user.getCuentasAhorrosAsociadas().get(i).getSaldo();
					contador = i;
				}
				Movimientos movimiento = new Movimientos(gota, user.getCuentasAhorrosAsociadas().get(contador),
						cantidadPrestamo, Categoria.OTROS, Date.from(Instant.now()));
				Movimientos.getMovimientosTotales().remove(movimiento);
			}
			return user.getCuentasAhorrosAsociadas().get(contador);
		} else {
			for (int i = 0; i < user.getCuentasCorrienteAsociadas().size(); i++) {
				if (user.getCuentasCorrienteAsociadas().get(i).getCupo() > mayor) {
					mayor = user.getCuentasCorrienteAsociadas().get(i).getCupo();
					contador = i;
				}
				Movimientos movimiento = new Movimientos(gota, user.getCuentasCorrienteAsociadas().get(contador),
						cantidadPrestamo, Categoria.OTROS, Date.from(Instant.now()));
				Movimientos.getMovimientosTotales().remove(movimiento);
			}
			return user.getCuentasCorrienteAsociadas().get(contador);
		}
	}
	
	// Métodos para funcionlidad de cambio de divisa
	public static void hacerCambio(Movimientos escogencia, double monto, Ahorros destino, Usuario user) {
		Ahorros origen = (Ahorros) escogencia.getOrigen(); 
		double cambiado = monto*(1-escogencia.getBanco().getEstadoAsociado().getTasa_impuestos());
		cambiado = cambiado*(1-escogencia.getCoutaManejo());
		cambiado = cambiado*(escogencia.getCantidad());
		Movimientos m = new Movimientos(escogencia.getBanco(), origen, destino, escogencia.getDivisa(), escogencia.getDivisaAux(), escogencia.getCoutaManejo() , monto, Date.from(Instant.now()));
		user.asociarMovimiento(m);
		origen.setSaldo(origen.getSaldo()-monto);
		destino.setSaldo(destino.getSaldo()+cambiado);
		for (int i = 0; i < user.getBancosAsociados().size() ; i++) {
			user.getBancosAsociados().get(i).setAsociado(false);
		}
	}
	
	public static void hacerCambio(Movimientos escogencia, double monto, Ahorros destino, boolean exacto, Usuario user) {
		Ahorros origen = (Ahorros) escogencia.getOrigen(); 
		double pagar = monto/(1-escogencia.getBanco().getEstadoAsociado().getTasa_impuestos());
		pagar = pagar/(1-escogencia.getCoutaManejo());
		pagar = pagar/(escogencia.getCantidad());
		Movimientos m = new Movimientos(escogencia.getBanco(), origen, destino, escogencia.getDivisa(), escogencia.getDivisaAux(), escogencia.getCoutaManejo() , pagar, Date.from(Instant.now()));
		user.asociarMovimiento(m);
		origen.setSaldo(origen.getSaldo()-pagar);
		destino.setSaldo(destino.getSaldo()+monto);
		for (int i = 0; i < user.getBancosAsociados().size() ; i++) {
			user.getBancosAsociados().get(i).setAsociado(false);
		}
	}
	
	public static ArrayList<Ahorros> obtenerCuentasDivisa(Usuario usuario, Divisas divisa){
		ArrayList<Ahorros> cuentasB = new ArrayList<Ahorros>();
		for (Ahorros ahorro: usuario.getCuentasAhorrosAsociadas()) {
			if (ahorro.getDivisa().equals(divisa)) {
				cuentasB.add(ahorro);
			}
		}
		return cuentasB;
	}
	
	public static boolean comprobarSaldo(Cuenta cuenta, double monto) {
		Ahorros ahorro = (Ahorros) cuenta;
		if (ahorro.getSaldo() >= monto) {
			return true;
		}
		else {
			return false;
		}
	}
	
	public static boolean comprobarSaldo(Movimientos escogencia, double monto ) {
		double pagar = monto/(1-escogencia.getBanco().getEstadoAsociado().getTasa_impuestos());
		pagar = pagar/(1-escogencia.getCoutaManejo());
		pagar = pagar/(escogencia.getCantidad());
		Ahorros ahorro = (Ahorros) escogencia.getOrigen();
		if (ahorro.getSaldo() >= monto) {
			return true;
		}
		else {
			return false;
		}
	}
		
//	public static ArrayList<Cuenta> cuentasConSaldo(ArrayList<Cuenta> cuentasDisponibes, double monto){
//		ArrayList<Cuenta> cuentasConSaldo = new ArrayList<Cuenta>(); 
//		for (Cuenta cuenta : ArrayList<Cuenta>; double monto) {
//			if (cuenta.getSaldo >= monto) {
//				cuentasConSaldo.add(cuenta);
//			}
//		}
//	}
	// Implementación métodos abstracto a redefinir
	public abstract void vaciarCuenta(Ahorros gota);
	
	public static double redondeoDecimal(double numero, int decimales) {
		double numRedondeado = Math.round(numero * Math.pow(10.0, decimales)) / Math.pow(10.0, decimales);
		return numRedondeado;
	}
	
	//Implementación de la interfaz Comparable
	public int compareTo(Cuenta cuenta) {
		if (this.getId() > cuenta.getId()) {
			return 1;
		}
		else if(this.getId() < cuenta.getId()) {
			return -1;
		}
		else {
			return 0;
		}
	}
	//Para ordenar cualquier arreglo de tipo Cuenta, se ordenará según el id de la cuenta y se hará con Collections.sort(nombre_lista);

	//	Ligadura dinamica
	public Object invertirSaldo(){
		float probabilidad = this.getTitular().getSuscripcion().getProbabilidad_Inversion();
		double rand = (double)((Math.random()) + probabilidad);
		if(this instanceof Ahorros){
			if(rand < 1){
				Ahorros cuenta = (Ahorros) this;
				return(Movimientos.crearMovimiento(cuenta,-(cuenta.getSaldo() - cuenta.getSaldo()*0.25),Categoria.FINANZAS,new Date())+"" +
						"\nLastimosamente debido a una mala inversion usted ha pedido un 25% de su saldo");
			}
		}
		return ("Esta cuenta no puede hacer inversiones");

	}

	public double[] retornoCuotaMensual(double DeudaActual) {
		Banco banco = this.getBanco();
		double[] cuotaMensual = new double[3];
		if(banco.getComision()+ banco.getEstadoAsociado().getInteres_bancario_corriente() <1){
//			Cuota del estado y del banco
			double cuota1 = DeudaActual*banco.getComision() + banco.getEstadoAsociado().getInteres_bancario_corriente()*DeudaActual;
			double cuota2 = (DeudaActual-cuota1)/2;
			cuotaMensual[0] = DeudaActual*banco.getComision() + banco.getEstadoAsociado().getInteres_bancario_corriente()*DeudaActual;
			cuotaMensual[1] = cuota2;
			cuotaMensual[1] = cuota2;
		}else{
			cuotaMensual[0] = DeudaActual/3;
			cuotaMensual[1] = DeudaActual/3;
			cuotaMensual[2] = DeudaActual/3;
		}
		return cuotaMensual;
	}

	@Override
	public boolean equals(Object o) {
		if(this.getId() == ((Cuenta) o).getId()){
			return true;
		}else {
			return false;
		}	
	}
	
	public static void limpiarPropiedades(ArrayList<String> arreglo){
		arreglo.remove("serialVersionUID");
		arreglo.remove("nombreD");
		arreglo.remove("cuentasTotales");
	}
	
	public static double DineroaTenerDisponible(Cuenta cuenta, Divisas divisaB) {
		//Tienes una cuenta de la que extraes su deuda (con el cupo y el disponible)
		//De la misma cuenta extraes la divisa en la que esta ese valor
		
		//Recibes por parámetro la divisa a la cual quiero pasar ese dinero
		
		//Haces tus cálculos y me devuelves cual es el valor que debo checar este disponible en la otra cuenta
		//El double debe resultar en la divisa que recibe el parámetro
		//Tu veras si cobras comisiones o no o si lo quieres traspasar a otro código
		Corriente corriente = (Corriente) cuenta;
		double deuda = (corriente.getCupo() - corriente.getDisponible());
		Movimientos cambioDiv = new Movimientos(cuenta.getDivisa() , divisaB, cuenta.getTitular());
		ArrayList<Banco> existeCambio = Movimientos.facilitarInformación(cambioDiv);
		if (existeCambio.size() == 0) {
			return 0;
		}
		ArrayList<Cuenta> cuentasPosibles = new ArrayList<Cuenta>();
		for (Cuenta conta : cuenta.getTitular().getCuentasCorrienteAsociadas()) {
			if (conta.getDivisa().equals(cuenta.getDivisa())) {
				cuentasPosibles.add(conta);
			}
		}
		String cadena = cuenta.getDivisa().name()+ divisaB.name();
		ArrayList<Movimientos> imprimir = Banco.cotizarTazaAux(cuenta.getTitular(), existeCambio, cadena, cuentasPosibles);
		double cambioMax = 0;
		double valor = 999999999;
		for (Movimientos cotizacion : imprimir) {
			valor = cotizacion.getCantidad()/((1-cotizacion.getCoutaManejo())*(1-cotizacion.getBanco().getEstadoAsociado().getTasa_impuestos()));
			if (valor > cambioMax) {
				cambioMax = valor;
			}
		}
		return cambioMax*deuda;
	}
	
	
	@Override	
	protected void finalize() {
		System.out.println("La cuenta " + this.getClass().getSimpleName() + " con id: " + this.getId() + " y nombre: " + this.getNombre() + " fue eliminada satisfactoriamente del sistema.");
	}

	public String toString() {
		return "Cuenta: " + this.nombre 
				+ "\n# " + this.id 
				+ "\nTitular: " + this.getTitular().getNombre() 
				+ "\nBanco: " + this.banco.getNombre() 
				+ "\nDivisa: " + this.divisa;
	}
	
	//Gets && Sets
	public static ArrayList<Cuenta> getCuentasTotales(){
		return Cuenta.cuentasTotales;
	}
	
	public static void setCuentasTotales(ArrayList<Cuenta> cuentasTotales){
		Cuenta.cuentasTotales = cuentasTotales;
	}
	
	public Usuario getTitular() {
		return titular;
	}
	
	public void setTitular(Usuario titular) {
		this.titular = titular;
	}
	
	public Divisas getDivisa() {
		return divisa;
	}
	
	public void setDivisa(Divisas divisa) {
		this.divisa = divisa;
	}
	
	public String getNombre() {
		return nombre;
	}
	
	public void setNombre(String nombre) {
		this.nombre = nombre;
	}
	
	public int getId() {
		return id;
	}
	
	public void setId(int id) {
		this.id = id;
	}
	
	public Banco getBanco() {
		return banco;
	}
	
	public void setBanco(Banco banco) {
		this.banco = banco;
	}

	public int getClave() {
		return clave;
	}
	
	public void setClave(int clave) {
		this.clave = clave;
	}
}
